### ISSUE TITLE
ORA-12704: character set mismatch in generated query

### ISSUE THREAD

**Issue by poke** [oracle/dotnet-db-samples#399]
Hi,

this issue might be related to #66, or the recent fixed bug “Bug 35535281 - DEFAULT VALUE FOR NON-UNICODE COLUMN IS GENERATED WITH N PREFIX WHEN USING VALUE CONVERTER AND ENUM TYPE“.

I am getting an ORA-12704 from the following query when applied on a non-Unicode string column:

```csharp
var results = db.Tests
    .Select(t => t.Title.Length > 80 ? t.Title.Substring(0, 77) + "..." : t.Title)
    .ToList();
```

This generates the following SQL:

```sql
SELECT CASE
    WHEN CAST(LENGTH("t"."TITLE") AS NUMBER(10)) > 80 THEN COALESCE(SUBSTR("t"."TITLE", 0 + 1, 77), NULL) || N'...'
    ELSE "t"."TITLE"
END
FROM "TEST" "t"
```

As you can see, it tries to append `N'...'` here even though the column in this case does not support unicode. The column is defined as `IsUnicode(false)`, so I understand it that it shouldn’t use the `N` prefix here.

If I drop out the substring logic, it also appends `N'...'` but then it works:

```csharp
var results = db.Tests
    .Select(t => t.Title + "...")
    .ToList();
```
```sql
SELECT COALESCE("t"."TITLE", NULL) || N'...'
FROM "TEST" "t"
```

Full test code is in the expandable section below.

<details>
  <summary>Show full test code</summary>

```sql
CREATE TABLE Test (
    Id NUMBER(19) GENERATED BY DEFAULT ON NULL AS IDENTITY NOT NULL,
    Title VARCHAR2(1024 CHAR),
    CONSTRAINT PK_Test PRIMARY KEY (Id)
)
```

```csharp
using Microsoft.EntityFrameworkCore;
using System;
using System.Linq;

var connectionString = "Data Source=…";

var dbOptions = new DbContextOptionsBuilder<MyContext>()
    .UseOracle(connectionString, oracleOptions =>
        oracleOptions.UseOracleSQLCompatibility(OracleSQLCompatibility.DatabaseVersion19))
    .Options;

using var db = new MyContext(dbOptions);

var results = db.Tests
    .Select(t => t.Title.Length > 80 ? t.Title.Substring(0, 77) + "..." : t.Title)
    //.Select(t => t.Title + "...")
    .ToList();

foreach (var result in results)
    Console.WriteLine(result);

public class MyContext : DbContext
{
    public DbSet<Test> Tests { get; set; }
    public MyContext(DbContextOptions options) : base(options) { }
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Test>(entity =>
        {
            entity.ToTable("TEST");
            entity.Property(e => e.Id).HasColumnName("ID").ValueGeneratedOnAdd();
            entity.Property(e => e.Title).HasColumnName("TITLE").HasMaxLength(1024).IsUnicode(false);
        });
    }
}
public class Test
{
    public long Id { get; set; }
    public string Title { get; set; }
}
```

</details>

I’ve tested this on Oracle.EntityFrameworkCore version 8.23.50. The target is an Oracle 19 database, which I have also set the compatibility level to.

**Comment by alexkeh** [comment]
@poke If you downgrade to Oracle EF Core 8.21.150, is the problem resolved? Do you see the "N" added in front of any string literal you use even if it's not "..."?

If so, this sounds like bug 36859898, which was earlier reported a couple weeks ago on the [ODP.NET OTN forum](https://forums.oracle.com/ords/apexds/post/8-25-x-oracle-entityframeworkcore-string-literals-now-causi-2188).

Let me know if the answer to both questions above is yes.

**Comment by poke** [comment]
@alexkeh Thanks for the quick response.

Downgrading to 8.21.150, the above code works and generates the following SQL which does not include the `N`:

```sql
SELECT CASE
    WHEN CAST(LENGTH("t"."TITLE") AS NUMBER(10)) > 80 THEN COALESCE(SUBSTR("t"."TITLE", 0 + 1, 77), NULL) || '...'
    ELSE "t"."TITLE"
END
FROM "TEST" "t"
```

So that’s the bug 36859898 that you have opened then? Would you suggest downgrading to 8.21 for the time being then? I guess that shouldn’t be a problem for us since we are targeting Oracle 19 anyway.

**Comment by poke** [comment]
Oh, and I missed this part of the question before:

> Do you see the "N" added in front of any string literal you use **even if it's not "..."**?

Yes, it seems to append a Unicode string regardless of what I do. It also does that in the simple (working) example where I don’t have a condition inside.

I guess that appending a column with a Unicode string will automatically convert the result into a Unicode result type. It’s just inside of the CASE…WHEN that it breaks when one branch is converted into Unicode while the other is not.

**Comment by alexkeh** [comment]
Yes, I opened this bug based on the test case filed in the OTN forum. I've noted in the bug log a second developer has hit the same issue and increased the priority to get a fix.

Yes, I would suggest downgrading to 21c until the bug is resolved.

**Comment by poke** [comment]
Alright, I’ll prepare for a downgrade then and keep an eye on the bug progress. Thanks a lot!

**Comment by poke** [comment]
I have to update this unfortunately with another finding:

Just downgrading to Oracle.EntityFrameworkCore 8.21.150 is not enough. In my test project, this worked correctly, but then failed in our actual application. After a lof of debugging, I found out that `Microsoft.EntityFrameworkCore.Relational` is also an important factor.

As soon as a version of EFCore.Relational is used above 8.0.2, the SQL translation produces a `N` prefix even with 8.21.150. If I pin EFCore.Relational to 8.0.1, the `N` prefix is gone and everything works.

```xml
<!-- This works -->
<PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="8.0.1" />
<PackageReference Include="Oracle.EntityFrameworkCore" Version="8.21.150" />

<!-- This does not -->
<PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="8.0.2" />
<PackageReference Include="Oracle.EntityFrameworkCore" Version="8.21.150" />
```

So maybe this is not an Oracle regression after all but related to the changes that EF Core introduced between 8.0.1 and 8.0.2. There are not many changes in that range affecting EFCore.Relational and I can spot two potential candidates that could be related to this issue:

- dotnet/efcore@fed6fe113699c923242f4751eee9dcdf60d523d9: “Infer type mapping suitable for concatenating two results”
- dotnet/efcore@93820b50a57244916d4e81d0923d2cba11f1d993: “Preserve unicodeness and fixed-lengthness in compiled model”

I don’t know enough about EF Core’s internals (and Oracle’s implementation of course) to know if this could have an effect but maybe that helps you resolve this issue better.

**Comment by poke** [comment]
> So maybe this is not an Oracle regression after all but related to the changes that EF Core introduced between 8.0.1 and 8.0.2.

For what it’s worth, Oracle.EntityFrameworkCore versions 8.23.xxx has a hard dependency on Microsoft.EntityFrameworkCore.Relational >= 8.0.3, so this could really be the issue here.

**Comment by alexkeh** [comment]
@poke The ODP.NET dev team performed some analysis in parallel and they have come to a similar conclusion. This bug is a result of the same root cause as https://github.com/oracle/dotnet-db-samples/issues/362. To summarize the findings, I'll quote myself:

"After some analysis, Oracle believes the bug is caused by EF Core relational layer changes in 8.0.2.

The MS code change was not expected in the EF Core relational layer, which introduced this regression.

Oracle raised a concern about the https://github.com/dotnet/efcore/pull/32520.

Meanwhile, as a workaround, developers that encounter this problem can set the below switch, which will reverse the changes the MS team has applied:

```
AppContext.SetSwitch("Microsoft.EntityFrameworkCore.Issue32325", true);

var list = new List<string>(){ "TEST1$$1", "TEST2$$1"};
var result = context.SampleTable.Where(a => list.Contains(a.Code1 + "$$" + a.Code2)).ToList();
```

MS Base bug: https://github.com/dotnet/efcore/issues/33218"

As such, we're waiting for MS to resolve the base regression. Oracle is also looking at ways it may be able to solve the problem in our own provider.

**Comment by poke** [comment]
@alexkeh Thanks a lot for the analysis and the information about that switch. That seems to be a more stable workaround than pinning 8.0.1.

Thanks also in general for the speedy support you're giving here. I really appreciate that! 😊

**Comment by alexkeh** [comment]
The ODP.NET team has implemented a workaround fix at the Oracle EF Core provider level. It will be available with version 23.6.
